src/npe_typedefs.h

```diff
@@ -247,6 +254,33 @@ enum TypeId {
   sparse_bool_x = 98,
 };
 
+// These ifs could probably be compile time.
+//
+// But in the end is this an elaborate way to change dtype().type() 'Q','q' -> 'L','l' ?
+//
+// Yes, but I guess the point is that these checks are what the code should be doing instead of using the chars to begin with.
```

Comment:
These characters are unique identifiers for the internal numpy types (https://numpy.org/doc/stable/reference/generated/numpy.dtype.char.html#numpy.dtype.char). 

If I remember correctly, we need a way of translating from the pybind array type into an identifier, and also being able to return something in the case where the array is something weird (like `Object`) which will throw an error.

I suppose you're right though, this could all be inlined into the generated code.